### PR TITLE
etcdserver: force backend commit before acking physical compaction

### LINKED ISSUE #4914: functional-tester: different hash after compaction
```
2016-03-30 23:40:14.676027 I | etcd-tester: [round#4] compacting storage at 721759 (current revision 731759)
2016-03-30 23:40:19.125544 I | etcd-tester: 10.240.0.3:2379 is already compacted with 721759 (rpc error: code = 11 desc = "etcdserver: storage: required revision has been compacted")
2016-03-30 23:40:19.167118 I | etcd-tester: 10.240.0.4:2379 is already compacted with 721759 (rpc error: code = 11 desc = "etcdserver: storage: required revision has been compacted")
2016-03-30 23:40:19.167192 I | etcd-tester: [round#4] compacted storage
2016-03-30 23:40:19.167206 I | etcd-tester: [round#4] check compaction at 721759
2016-03-30 23:40:19.268869 I | etcd-tester: [round#4] confirmed compaction at 721759
2016-03-30 23:40:19.419253 I | etcd-tester: [round#5 case#0] start failure kill all members
2016-03-30 23:40:19.419289 I | etcd-tester: [round#5 case#0] start injecting failure...
2016-03-30 23:40:19.483197 I | etcd-tester: [round#5 case#0] injected failure
2016-03-30 23:40:19.483234 I | etcd-tester: [round#5 case#0] start recovering failure...
2016-03-30 23:40:22.165048 I | etcd-tester: [round#5 case#0] recovered failure
2016-03-30 23:40:22.165099 I | etcd-tester: [round#5 case#0] canceling the stressers...
2016-03-30 23:40:22.166190 I | etcd-tester: [round#5 case#0] canceled stressers
2016-03-30 23:40:22.166217 I | etcd-tester: [round#5 case#0] checking current revisions...
2016-03-30 23:40:23.280842 I | etcd-tester: [round#5 case#0] all members are consistent with current revisions
2016-03-30 23:40:23.280894 I | etcd-tester: [round#5 case#0] checking current storage hashes...
2016-03-30 23:40:23.280936 I | etcd-tester: [round#5 case#0] checking current storage hashes failed (map[10.240.0.2:2379:403078244 10.240.0.3:2379:403078244 10.240.0.4:2379:2058040325])
2016-03-30 23:40:23.280961 I | etcd-tester: [round#5 case#0] cleaning up...
```

### PR DESCRIPTION
**Pull request by heyitsanthony**
Fixes #4914

### CODE REVIEW

**Comment by heyitsanthony** [comment]
/cc @gyuho @xiang90 

Not sure about moving the `go s.Compact()`, but otherwise there's a race on the compaction scheduler queue.

**Comment by gyuho** [comment]
lgtm

**Comment by gyuho** [comment]
just deployed with this patch. will check when home.

thanks!

**Comment by gyuho** [comment]
Still seeing errors. I will check the logs.

File: etcdserver/v3demo_server.go [hunk 1]
```diff
@@ -96,6 +96,8 @@ func (s *EtcdServer) Compact(ctx context.Context, r *pb.CompactionRequest) (*pb.
 	result, err := s.processInternalRaftRequest(ctx, pb.InternalRaftRequest{Compaction: r})
 	if r.Physical && result.physc != nil {
 		<-result.physc
+		// if server goes down, should not replay the compaction
+		s.be.ForceCommit()
```
Comment (xiang90) [inline etcdserver/v3demo_server.go]: i do not quite understand the relationship between the comment and the code. can we be more clear?

File: storage/kvstore.go [hunk 2]
```diff
@@ -277,6 +277,11 @@ func (s *store) Compact(rev int64) (<-chan struct{}, error) {
 
 func (s *store) Hash() (uint32, error) {
 	s.b.ForceCommit()
+	s.mu.Lock()
```
Comment (xiang90) [inline storage/kvstore.go]: i think the test code first calls compact and waits for its finish, then call Hash(). So will this actually make a change?

**Comment by xiang90** [comment]
@gyuho Have you tried the new change again?

**Comment by gyuho** [comment]
@xiang90 I just saw this new commit. I will deploy this right now. Will let you know.

**Comment by gyuho** [comment]
```
# TESTER
2016-04-01 02:27:16.776698 I | etcd-tester: [round#4] compacting storage at 915405 (current revision 925405)
2016-04-01 02:27:16.776725 I | etcd-tester: compactKV to 10.240.0.2:2379
2016-04-01 02:27:21.782445 I | etcd-tester: compactKV to 10.240.0.3:2379
2016-04-01 02:27:21.790789 I | etcd-tester: 10.240.0.3:2379 is already compacted with 915405 (rpc error: code = 11 desc = "etcdserver: storage: required revision has been compacted")
2016-04-01 02:27:21.790836 I | etcd-tester: [round#4] compacted storage
2016-04-01 02:27:21.790845 I | etcd-tester: [round#4] check compaction at 915405
2016-04-01 02:27:22.050951 I | etcd-tester: [round#4] confirmed compaction at 915405
2016-04-01 02:27:22.149278 I | etcd-tester: [round#5 case#0] start failure kill all members
2016-04-01 02:27:22.149311 I | etcd-tester: [round#5 case#0] start injecting failure...
2016-04-01 02:27:22.195161 I | etcd-tester: [round#5 case#0] injected failure
2016-04-01 02:27:22.195205 I | etcd-tester: [round#5 case#0] start recovering failure...
2016-04-01 02:27:25.014688 I | etcd-tester: [round#5 case#0] recovered failure
2016-04-01 02:27:25.014752 I | etcd-tester: [round#5 case#0] canceling the stressers...
2016-04-01 02:27:25.016830 I | etcd-tester: [round#5 case#0] canceled stressers
2016-04-01 02:27:25.016864 I | etcd-tester: [round#5 case#0] checking current revisions...
2016-04-01 02:27:26.139656 I | etcd-tester: [round#5 case#0] all members are consistent with current revisions
2016-04-01 02:27:26.139688 I | etcd-tester: [round#5 case#0] checking current storage hashes...
2016-04-01 02:27:26.139717 I | etcd-tester: [round#5 case#0] checking current storage hashes failed (map[10.240.0.2:2379:491388969 10.240.0.3:2379:3650728422 10.240.0.4:2379:4245303024])
2016-04-01 02:27:26.139751 I | etcd-tester: [round#5 case#0] cleaning up...

# AGENT1
2016-04-01 02:27:16.744701 I | store.index: compact 915405
2016-04-01 02:27:16.957487273 +0000 UTC scheduleCompaction started at 915405
2016-04-01 02:27:23.627230 I | store.index: compact 915405
2016-04-01 02:27:23.794842 I | storage: resume scheduled compaction at 915405
2016-04-01 02:27:23.795069971 +0000 UTC scheduleCompaction started at 915405

# AGENT2
2016-04-01 02:27:16.805955 I | store.index: compact 915405
2016-04-01 02:27:17.021034993 +0000 UTC scheduleCompaction started at 915405
2016-04-01 02:27:23.920387 I | store.index: compact 915405
2016-04-01 02:27:24.062318 I | storage: resume scheduled compaction at 915405
2016-04-01 02:27:24.06253477 +0000 UTC scheduleCompaction started at 915405

# AGENT3
2016-04-01 02:27:16.775875 I | store.index: compact 915405
2016-04-01 02:27:16.931176787 +0000 UTC scheduleCompaction started at 915405
2016-04-01 02:27:23.527015 I | store.index: compact 915405
2016-04-01 02:27:23.731847 I | storage: resume scheduled compaction at 915405
2016-04-01 02:27:23.732032977 +0000 UTC scheduleCompaction started at 915405
2016-04-01 02:27:26.302388449 +0000 UTC scheduleCompaction done at 915405
```


with https://github.com/gyuho/etcd/commit/6925a4391b91ce886d8906fe3ad8d05a1f3845d6, which is same as this patch plus more logging lines. `checkCompact` still returns before compaction operations are finished:

```
2016-04-01 02:27:22.050951 I | etcd-tester: [round#4] confirmed compaction at 915405

2016-04-01 02:27:23.795069971 +0000 UTC scheduleCompaction started at 915405
2016-04-01 02:27:24.06253477 +0000 UTC scheduleCompaction started at 915405
2016-04-01 02:27:26.302388449 +0000 UTC scheduleCompaction done at 915405
```

Any chance that we update `compactRev` in watch response that makes `checkCompact` function believe the compaction is done?

**Comment by heyitsanthony** [comment]
@gyuho https://github.com/gyuho/etcd/blob/test/tools/functional-tester/etcd-tester/cluster.go#L341-L347 isn't using `compactKV` like a barrier. The RPC has to be sent to every node to know they're all physically synced.

**Comment by gyuho** [comment]
@heyitsanthony Right. I will try again sending request to every node. Thanks!

**Comment by heyitsanthony** [comment]
@gyuho what's in #4910 would be good

**Comment by gyuho** [comment]
yeah just redeployed. I will keep updates here. Thanks!

**Comment by gyuho** [comment]
@heyitsanthony Great. Think this fixes the issue. We can merge?

**Comment by heyitsanthony** [comment]
fixed up the comment about the forcecommit, merging on greenlight

**Comment by gyuho** [comment]
@heyitsanthony We don't need etcdserver change for retrial?

**Pull request merged**